File: backend/src/domain/entities/site.py
# backend/src/domain/entities/site.py
from __future__ import annotations
from dataclasses import dataclass, field
from datetime import date, datetime, timezone
from enum import StrEnum
from typing import Optional


class SiteStatus(StrEnum):
    INSTALLING  = "installing"
    ACTIVE      = "active"
    INACTIVE    = "inactive"
    EXPIRED     = "expired"
    MAINTENANCE = "maintenance"


class ContractType(StrEnum):
    OFFICIAL  = "정식라이센스"
    TEMPORARY = "임시라이센스"


class NodeRole(StrEnum):
    ALL_IN_ONE = "AllInOne"
    ANALYZER   = "Analyzer"
    COLLECTOR  = "Collector"
    PROXY      = "Proxy"


class PatchType(StrEnum):
    REGULAR   = "정기패치"
    EMERGENCY = "긴급패치"
    HOTFIX    = "핫픽스"


class PatchResultStatus(StrEnum):
    SUCCESS     = "성공"
    FAILED      = "실패"
    ROLLED_BACK = "롤백"


def _utc_now() -> datetime:
    return datetime.now(timezone.utc)


@dataclass(frozen=True)
class ContactInfo:
    name:    Optional[str] = None
    phone:   Optional[str] = None
    email:   Optional[str] = None
    company: Optional[str] = None


@dataclass(frozen=True)
class Credential:
    username: str
    password: str
    ip:       Optional[str] = None
    port:     Optional[str] = None


@dataclass(frozen=True)
class AccessCredentials:
    cli:  Optional[Credential] = None
    web:  Optional[Credential] = None
    db:   Optional[Credential] = None
    vpn:  Optional[Credential] = None
    note: Optional[str] = None


@dataclass(frozen=True)
class DeploymentNode:
    id:              Optional[int]      = None
    hostname:        Optional[str]      = None
    role:            Optional[NodeRole] = None
    cpu_cores:       Optional[int]      = None
    cpu_threads:     Optional[int]      = None
    memory_total_gb: Optional[int]      = None
    disk_total_gb:   Optional[int]      = None
    os_type:         Optional[str]      = None
    os_version:      Optional[str]      = None
    ip_address:      Optional[str]      = None
    disk_free_gb:    Optional[int]      = None
    disk_updated_at: Optional[datetime] = None
    pkg_version:     Optional[str]      = None


@dataclass(frozen=True)
class PatchHistory:
    id:              Optional[int]               = None
    issue_link:      Optional[str]               = None
    patch_date:      Optional[date]              = None
    patch_file_link: Optional[str]               = None
    patch_type:      Optional[PatchType]         = None
    applied_by:      Optional[str]               = None
    result_status:   Optional[PatchResultStatus] = None
    rollback_date:   Optional[date]              = None
    note:            Optional[str]               = None


@dataclass(frozen=True)
class VisitHistory:
    id:               Optional[int]      = None
    visit_datetime:   Optional[datetime] = None
    engineer_name:    Optional[str]      = None
    engineer_phone:   Optional[str]      = None
    request_content:  Optional[str]      = None
    action_content:   Optional[str]      = None


@dataclass
class Site:
    site_name:           str
    id:                  Optional[int]               = None
    maintenance_company: Optional[str]               = None
    customer_contact:    Optional[ContactInfo]       = None
    maintenance_contact: Optional[ContactInfo]       = None
    contract_start_date: Optional[date]              = None
    contract_end_date:   Optional[date]              = None
    contract_type:       Optional[ContractType]      = None
    status:              Optional[SiteStatus]        = None
    nodes:               list[DeploymentNode]        = field(default_factory=list)
    patch_histories:     list[PatchHistory]          = field(default_factory=list)
    visit_histories:     list[VisitHistory]          = field(default_factory=list)
    access_credentials:  Optional[AccessCredentials] = None
    created_at:          datetime                    = field(default_factory=_utc_now)
    updated_at:          datetime                    = field(default_factory=_utc_now)


@dataclass(frozen=True)
class SiteSummary:
    id:                int
    site_name:         str
    customer_name:     Optional[str] = None
    status:            Optional[SiteStatus] = None
    contract_end_date: Optional[date] = None

File: frontend/src/domain/Site.ts
// frontend/src/domain/Site.ts
export type SiteStatus = 'installing' | 'active' | 'inactive' | 'expired' | 'maintenance'

export type ContractType = '정식라이센스' | '임시라이센스'

export type NodeRole = 'AllInOne' | 'Analyzer' | 'Collector' | 'Proxy'

export type PatchType = '정기패치' | '긴급패치' | '핫픽스'

export type PatchResultStatus = '성공' | '실패' | '롤백'

export interface ContactInfo {
  name?: string
  phone?: string
  email?: string
  company?: string
}

export interface Credential {
  username?: string
  password?: string
  ip?: string
  port?: string
}

export interface AccessCredentials {
  cli?: Credential
  web?: Credential
  db?: Credential
  vpn?: Credential
  note?: string
}

export interface DeploymentNodePayload {
  hostname?: string
  role?: NodeRole
  cpu_cores?: number
  cpu_threads?: number
  memory_total_gb?: number
  disk_total_gb?: number
  os_type?: string
  os_version?: string
  ip_address?: string
  disk_free_gb?: number
  disk_updated_at?: string
  pkg_version?: string
}

export interface DeploymentNode extends DeploymentNodePayload {
  id?: number
}

export interface PatchHistoryPayload {
  patch_date?: string
  patch_type?: PatchType
  applied_by?: string
  result_status?: PatchResultStatus
  issue_link?: string
  patch_file_link?: string
  rollback_date?: string
  note?: string
}

export interface PatchHistory extends PatchHistoryPayload {
  id?: number
}

export interface VisitHistoryPayload {
  visit_datetime?: string
  engineer_name?: string
  engineer_phone?: string
  request_content?: string
  action_content?: string
}

export interface VisitHistory extends VisitHistoryPayload {
  id?: number
}

export interface SiteSummary {
  id: number
  site_name: string
  customer_name?: string
  status?: SiteStatus
  contract_end_date?: string
}

export interface SiteDetail {
  id: number
  site_name: string
  maintenance_company?: string
  customer_info?: ContactInfo
  maintenance_info?: ContactInfo
  contract_start_date?: string
  contract_end_date?: string
  contract_type?: ContractType
  status?: SiteStatus
  created_at: string
  updated_at: string
  nodes: DeploymentNode[]
  patch_histories: PatchHistory[]
  visit_histories: VisitHistory[]
  access_credentials?: AccessCredentials
}

export interface SiteCreatePayload {
  site_name: string
  maintenance_company?: string
  customer_info?: ContactInfo
  maintenance_info?: ContactInfo
  contract_start_date?: string
  contract_end_date?: string
  contract_type?: ContractType
  status?: SiteStatus
  nodes: DeploymentNodePayload[]
  patch_histories: PatchHistoryPayload[]
  visit_histories: VisitHistoryPayload[]
  access_credentials?: AccessCredentials
}

export interface SiteUpdatePayload {
  site_name?: string
  maintenance_company?: string
  customer_info?: ContactInfo
  maintenance_info?: ContactInfo
  contract_start_date?: string
  contract_end_date?: string
  contract_type?: ContractType
  status?: SiteStatus
  access_credentials?: AccessCredentials
}
